### PR TITLE
Fix composite compliance problems for linalg.{matrix_power, inv, cholesky}

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#69437 Fix composite compliance problems for linalg.{matrix_power, inv, cholesky}**

linalg.{inv, cholesky} are problematic because they call .data_ptr().
This makes them not composite compliant (e.g. meta tensors will not run
on them correctly). This PR makes them composite compliant by adding a
new native_functions operator that does error checking,
`_linalg_check_errors(Tensor info, str api_name, bool is_matrix`
that is a primitive with respect to autograd.

This PR modifies linalg.inv and linalg.cholesky to call the new error
check function. I also needed to refactor singleCheckErrors and
batchCheckErrors to accept a c10::string_view instead of a
`const char*`; you can convert `const char*` to c10::string_view but not
the other way around because `string_view` does not require null
terminated buffers.

Finally, there is a bugfix in `__torch_dispatch__` for this PR for
the composite compliant testing mechanism. Previously,
`__torch_dispatch__` could not handle operators with no returns; this PR
fixes that. No returns in C++ is equivalent to a single None return in
Python.

Test Plan:
- composite compliant tests

Differential Revision: [D32883666](https://our.internmc.facebook.com/intern/diff/D32883666)

### CODE REVIEW

**Comment by zou3519** [comment]
cc @Lezcano and @IvanKobzarev on the linalg changes. This is what we discussed in https://github.com/pytorch/pytorch/pull/69263

cc @albanD on the __torch_dispatch__ changes

**Review by lezcano — DISMISSED** [review]
Loving this PR! Thank you so much for this!

It would be great to change all the uses of the `*CheckErrors` to this new function, and it should be as straightforward as a find/replace. Feel free to do so, or just land this and we can do that in a separate follow-up PR.

File: torch/csrc/autograd/python_variable.cpp [hunk 1]
```diff
@@ -1785,7 +1785,11 @@ void concrete_dispatch_fn(
     "__torch_dispatch__"
   ));
 
-  if (op.schema().returns().size() == 1) {
+  if (num_returns == 0) {
```
Comment (albanD) [inline torch/csrc/autograd/python_variable.cpp:1788]: Good catch!

**Review by albanD — DISMISSED** [review]
SGTM thanks!

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32883666).

**Comment by zou3519** [comment]
> It would be great to change all the uses of the `*CheckErrors` to this new function, and it should be as straightforward as a find/replace. Feel free to do so, or just land this and we can do that in a separate follow-up PR.

I'm going to land this first to unblock functorch but I'm happy to do the follow-up on top of this

**Review by IvanYashchuk — DISMISSED** [review]

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -735,6 +735,17 @@ template<> void blasTriangularSolve<float>(char side, char uplo, char trans, cha
...
+void _linalg_check_errors(
+    const Tensor& info,
+    const c10::string_view api_name,
+    bool is_matrix) {
+  if (is_matrix) {
+    singleCheckErrors(info.item<int64_t>(), api_name);
+  } else {
+    batchCheckErrors(info, api_name);
+  }
+}
```
Comment (zou3519) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:747]: On second thought... does `is_matrix` need to be an argument to this function? Could we infer `is_matrix` from `info.numel() == 1`? cc @Lezcano @IvanYashchuk

Comment (Lezcano) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:747]: We can't do that because matrices of size `(1, 1, 3, 2)` and `(3, 2)` would be treated the same.
Now, we could probably infer from `info.dim() == 0`. What do you think @IvanYashchuk ?

Comment (IvanYashchuk) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:747]: I don't know if `info` is consistently implemented to have shape `()` for single input matrices, it could also be `(1,)`.

Comment (zou3519) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:747]: @Lezcano @IvanYashchuk another question: is the condition always "is a matrix"? Looking at some code (I decided to change usages of batchCheckErrors to this new logic in this PR to avoid churn), there are some cases where the condition is "result.dim() > 1":

https://github.com/pytorch/pytorch/blob/9aa1b3e396c1fe9dadecaee04c9321f659d863ac/aten/src/ATen/native/BatchLinearAlgebra.cpp#L962-L976

Should I rename (and invert) `is_matrix` to `is_batch` instead?

Comment (IvanYashchuk) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:747]: @zou3519, thank you for pointing out this use case that deviates from other uses of `info` to check the error! It was necessary here because input could be a single matrix but then internally it was expanded to be batched which affects the size of `infos`.

I submitted a pull request (https://github.com/pytorch/pytorch/pull/69752) fixing that and now the condition is `if (input.dim() > 2)` instead of `if (vector_case ? result.dim() > 1 : result.dim() > 2)`.

> Should I rename (and invert) is_matrix to is_batch instead?

I don't think it's necessary, only if you prefer "is_batch".

**Comment by Lezcano** [comment]
What's the state of this PR @zou3519 ?

**Comment by zou3519** [comment]
> What's the state of this PR @zou3519 ?

Oops, I'll clean this up and submit an update early next week

**Comment by zou3519** [comment]
This is ready for re-review. What changed since the last iteration is that I incorporated @Lezcano's suggestion to "change all the uses of the *CheckErrors to this new function".

NB: I wasn't able to get rid of all the batchCheckErrors callsites. There are three callsites that that use `vector<int64_t> infos` instead of Tensor infos
- https://github.com/pytorch/pytorch/blob/2bb6a4f437cab6bc1b57807a5bfa75c4f83f59be/aten/src/ATen/native/BatchLinearAlgebra.cpp#L1267-L1275
- https://github.com/pytorch/pytorch/blob/2bb6a4f437cab6bc1b57807a5bfa75c4f83f59be/aten/src/ATen/native/BatchLinearAlgebra.cpp#L2525-L2529
- https://github.com/pytorch/pytorch/blob/2bb6a4f437cab6bc1b57807a5bfa75c4f83f59be/aten/src/ATen/native/BatchLinearAlgebra.cpp#L3068-L3072

None of these matter for composite compliance though since they occur in non-composite functions.

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 3]
```diff
@@ -1176,14 +1167,8 @@ Tensor& linalg_inv_out(const Tensor &input, Tensor &result) {
...
-  if (result.dim() > 2) {
-    batchCheckErrors(infos_lu, "linalg.inv");
-    batchCheckErrors(infos_getri, "linalg.inv");
-  } else {
-    singleCheckErrors(infos_lu.item().toInt(), "linalg.inv");
-    singleCheckErrors(infos_getri.item().toInt(), "linalg.inv");
-  }
-
+  at::_linalg_check_errors(infos_lu, "linalg.inv", result.dim() == 2);
+  at::_linalg_check_errors(infos_getri, "linalg.inv", result.dim() == 2);
```
Comment (Lezcano) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:1171]: [Unrelated to this PR] @IvanYashchuk @nikitaved We should just check the infos of LU here, as if they are fine, getri should always succeed.

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -6945,6 +6945,10 @@
   structured_delegate: triangular_solve.X
   variants: method, function
 
+- func: _linalg_check_errors(Tensor info, str api_name, *, bool is_matrix) -> ()
+  dispatch:
+    CompositeExplicitAutograd: _linalg_check_errors
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:6950]: Why is this needed?

**Review by albanD — APPROVED** [review]
SGTM

File: aten/src/ATen/native/native_functions.yaml (hunk 4 above)
Comment (zou3519) [reply aten/src/ATen/native/native_functions.yaml:6950]: We're hiding the data_ptr calls in a new non-composite operator, `_linalg_check_errors`. Does that answer the question or were you referring to something specific in these three lines?

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32883666).

File: aten/src/ATen/native/native_functions.yaml (hunk 4 above)
Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml:6950]: Ho I was was refering to making this CompositeExplicitAutograd vs the default CompositeImplicitAutograd

Comment (zou3519) [reply aten/src/ATen/native/native_functions.yaml:6950]: Making this CompositeImplicitAutograd makes it so that this operation is composite with respect to autograd, which is a problem because internally it calls .data_ptr().

Making this CompositeExplicitAutograd makes it so that this operation is not composite with respect to autograd (I think? The naming is really weird) so that the .data_ptr() call is hidden.

Does that sound right?

Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml:6950]: Sounds good.

Comment (zou3519) [reply aten/src/ATen/native/native_functions.yaml:6950]: Side note: can we rename CompositeExplicitAutograd? I'm not sure what it's supposed to mean

Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml:6950]: This is the 3rd name for it already....
The idea is:
- CompositeImpliciteAutograd: is a composite op (used by all backends) that is autodiffed (has an implicit autograd).
- CompositeExplicitAutograd: is a composite op (used by all backends) that has an explicit autograd formula.

Comment (zou3519) [reply aten/src/ATen/native/native_functions.yaml:6950]: I see, so CompositeExplicitAutograd means it's composite w.r.t. to the backend, not that it is composite w.r.t autograd :P

Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml:6950]: Yes, the composite part is about aten.
implicit/explicit is about autograd.

We did discussed these names at length with Ailing. But it is very very hard to find a name that makes everyone happy :(

**Comment by zou3519** [comment]
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32883666).

**Review by IvanYashchuk — APPROVED** [review]
